src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/DefaultRazorTagHelperContextDiscoveryPhase.cs

```diff
@@ -106,32 +108,34 @@ private static ReadOnlySpan<char> RemoveGlobalPrefix(in ReadOnlySpan<char> span)
         return span;
     }
 
-    internal abstract class DirectiveVisitor(HashSet<TagHelperDescriptor> matches) : SyntaxWalker
+    internal abstract class DirectiveVisitor(HashSet<TagHelperDescriptor> matches) : SyntaxWalker, IDisposable
     {
         protected readonly HashSet<TagHelperDescriptor> Matches = matches;
 
         public abstract string TagHelperPrefix { get; }
 
         public abstract void Visit(RazorSyntaxTree tree);
+
+        public abstract void Dispose();
     }
 
     internal sealed class TagHelperDirectiveVisitor : DirectiveVisitor
     {
-        private readonly List<TagHelperDescriptor> _tagHelpers;
+        private readonly PooledObject<ImmutableArray<TagHelperDescriptor>.Builder> _tagHelpers;
```

Comment:
Random thought: Is ImmutableArray the best choice here? Since this builder is never realized, a pooled list might be just as suitable, but a bit simpler, especially if an initial capacity can be set?

Looks like `ImmutableArray<>.Builder` [just resizes an array](https://source.dot.net/#System.Collections.Immutable/System/Collections/Immutable/ImmutableArray_1.Builder.cs,741) anyway, so I can't imagine List would be any worse, but I don't really have any knowledge here.

Reply:
Or even just a `PooledArrayBuilder<T>`?

Reply:
I tried PooledArrayBuilder initially, and it didn't seem to work because I was trying to store that into a field, and that type is defined as a NonCopyable struct.

Reply:
I didn't go with PooledList as I talked to Andrew and we weren't sure of it's benefit. It's a struct, but not marked as non-copyable, so it could be used, but it doesn't seem to be used a lot throughout the codebase. I'm flexible, and can go with that (or whatever else works) if you'd prefer.

Reply:
Yeah, `PooledList` can't be used, but we also have [`ListPool`](https://github.com/dotnet/razor/blob/cd4bbd166a3c0cd4144eb5c7d3979e2b4cfec505/src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/PooledObjects/ListPool%601.cs#L17) which can. Isn't coding fun?!

Reply:
Yeah, this feels cleaner. Thanks for the suggestions!